lib/types/jtd-schema.ts

```diff
@@ -48,94 +75,87 @@ export type JTDSchemaType<T, D extends Record<string, unknown> = Record<string,
-      // strings - only accepts the type string
-      | (true extends TypeEquality<Exclude<T, null>, string> ? {type: StringType} : never)
-      // strings - only accepts the type Date
-      | (true extends TypeEquality<Exclude<T, null>, Date> ? {type: "timestamp"} : never)
-      // enums - only accepts union of string literals
+  | ((// numbers - only accepts the type number
+    true extends NullTypeEquality<T, number>
+      ? {type: NumberType}
+      : // booleans - accepts the type boolean
+      true extends NullTypeEquality<T, boolean>
+      ? {type: "boolean"}
+      : // strings - only accepts the type string
+      true extends NullTypeEquality<T, string>
+      ? {type: StringType}
+      : // strings - only accepts the type Date
+      true extends NullTypeEquality<T, Date>
+      ? {type: "timestamp"}
+      : // enums - only accepts union of string literals
       // TODO we can't actually check that everything in the union was specified
-      | (null extends EnumString<Exclude<T, null>> ? never : {enum: EnumString<Exclude<T, null>>[]})
-      // arrays - only accepts arrays, could be array of unions to be resolved later
-      | (false extends IsUnion<Exclude<T, null>>
-          ? true extends IsElements<Exclude<T, null>>
-            ? T extends readonly (infer E)[]
-              ? {
-                  elements: JTDSchemaType<E, D>
+      true extends IsEnum<T>
+      ? {enum: EnumString<Exclude<T, null>>[]}
+      : // arrays - only accepts arrays, could be array of unions to be resolved later
+      true extends IsElements<Exclude<T, null>>
+      ? T extends readonly (infer E)[]
+        ? {
+            elements: JTDSchemaType<E, D>
+          }
+        : never
+      : // values
+      true extends IsValues<T>
+      ? T extends Record<string, infer V>
+        ? {
+            values: JTDSchemaType<V>
+          }
+        : never
+      : // properties
+      true extends IsProperties<T>
+      ? ([RequiredKeys<Exclude<T, null>>] extends [never]
+          ? {
+              properties?: Record<string, never>
+            }
+          : {
+              properties: {[K in RequiredKeys<T>]: JTDSchemaType<T[K], D>}
+            }) &
+          ([OptionalKeys<Exclude<T, null>>] extends [never]
+            ? {
+                optionalProperties?: Record<string, never>
+              }
+            : {
+                optionalProperties: {
+                  [K in OptionalKeys<T>]: JTDSchemaType<Exclude<T[K], undefined>, D>
                 }
-              : never
-            : never
-          : never)
-      // values
-      | (false extends IsUnion<Exclude<T, null>>
-          ? true extends TypeEquality<keyof Exclude<T, null>, string>
-            ? T extends Record<string, infer V>
-              ? {
-                  values: JTDSchemaType<V>
+              }) & {
+            additionalProperties?: boolean
+          }
+      : // discriminator
+      true extends IsDiscriminator<T>
```

Comment:
... and this `IsRecord<T, true>`

Or they could stay separate, defined via `IsRecord`.